soh/soh/Enhancements/randomizer/hook_handlers.cpp

```diff
@@ -1483,6 +1574,284 @@ void RandomizerOnActorInitHandler(void* actorRef) {
+                currentBossSoulRandInf = RAND_INF_VOLVAGIA_SOUL;
+                break;
+            case SCENE_WATER_TEMPLE_BOSS:
+                currentBossSoulRandInf = RAND_INF_MORPHA_SOUL;
+                break;
+            case SCENE_SHADOW_TEMPLE_BOSS:
+                currentBossSoulRandInf = RAND_INF_BONGO_BONGO_SOUL;
+                break;
+            case SCENE_SPIRIT_TEMPLE_BOSS:
+                currentBossSoulRandInf = RAND_INF_TWINROVA_SOUL;
+                break;
+            case SCENE_GANONDORF_BOSS:
+            case SCENE_GANON_BOSS:
+                if (RAND_GET_OPTION(RSK_SHUFFLE_BOSS_SOULS) == RO_BOSS_SOULS_ON_PLUS_GANON) {
+                    currentBossSoulRandInf = RAND_INF_GANON_SOUL;
+                }
+                break;
+            default:
+                break;
+        }
+
+        //Deletes all actors in the boss category if the soul isn't found.
+        //Some actors, like Dark Link, Arwings, and Zora's Sapphire...?, are in this category despite not being actual bosses,
+        //so ignore any "boss" if `currentBossSoulRandInf` doesn't change from RAND_INF_MAX.
+        if (currentBossSoulRandInf != RAND_INF_MAX) {
+            if (!Flags_GetRandomizerInf(currentBossSoulRandInf) && actor->category == ACTORCAT_BOSS) {
+                Actor_Delete(&gPlayState->actorCtx, actor, gPlayState);
+            }
+            //Special case for Phantom Ganon's horse (and fake), as they're considered "background actors",
+            //but still control the boss fight flow.
+            if (!Flags_GetRandomizerInf(RAND_INF_PHANTOM_GANON_SOUL) && actor->id == ACTOR_EN_FHG) {
+                Actor_Delete(&gPlayState->actorCtx, actor, gPlayState);
+            }
+        }
+    }
+
+    // In MQ Spirit, remove the large silver block in the hole as child so the chest in the silver block hallway
+    // can be guaranteed accessible
+    if (
+        actor->id == ACTOR_OBJ_OSHIHIKI &&
+        LINK_IS_CHILD &&
+        IsGameMasterQuest() &&
+        gPlayState->sceneNum == SCENE_SPIRIT_TEMPLE && actor->room == 6 && // Spirit Temple silver block hallway
+        actor->params == 0x9C7 // Silver block that is marked as in the hole
+    ) {
+        Actor_Kill(actor);
+        return;
+    }
+
+    if (
+        // If child is in the adult shooting gallery or adult in the child shooting gallery, then despawn the shooting gallery man
+        actor->id == ACTOR_EN_SYATEKI_MAN &&
+        RAND_GET_OPTION(RSK_SHUFFLE_INTERIOR_ENTRANCES) &&
+        (
+            (LINK_IS_CHILD && Entrance_SceneAndSpawnAre(SCENE_SHOOTING_GALLERY, 0x00)) || //Kakariko Village -> Adult Shooting Gallery, index 003B in the entrance table
+            (LINK_IS_ADULT && Entrance_SceneAndSpawnAre(SCENE_SHOOTING_GALLERY, 0x01)) //Market -> Child Shooting Gallery,              index 016D in the entrance table
+        )
+    ) {
+        Actor_Kill(actor);
+        return;
+    }
+}
+
+void RandomizerOnGameFrameUpdateHandler() {
```

Comment:
Similarly here, we should only register this hook handler when the infinite upgrades option is enabled in the current save